editor/modules/runner/game-state.ts

```diff
@@ -0,0 +1,102 @@
+import * as sortedIndexBy from 'lodash/sortedIndexBy';
+
+import {Injectable} from '@angular/core';
+import {Observable} from 'rxjs/Observable';
+import {BehaviorSubject} from 'rxjs/BehaviorSubject';
+
+import {Vec2, length2} from '../../util/math';
+import {LycanService} from './lycan.service';
+import {Player} from './player';
+
+
+interface Entity {
+    pos: Vec2;
+    id: number;
+    next_pos: Vec2;
+    nominal_speed: number;
+    speed: Vec2;
+}
+
+
+@Injectable()
+export class GameState {
+
+    entities: Entity[] = [];
+
+    // The player entity
+    player: Entity;
+
+    private count_changed = new BehaviorSubject<Entity[]>(this.entities);
+
+    constructor(
+        private lycan: LycanService,
+        private player_info: Player
+    ) {
+        this.lycan.getUpdateStream().subscribe(up => {
+            switch (up.kind) {
+                case 'NewEntity':
```

Comment:
This act exactly like Rust enum (sum type). The `kind` field act like a discriminant. Everything is type-checked. If you play with it, and replace `'NewEntity'` with `'newEntity'` you'll get an error. If you replace it with another valid variant you'll get an error within the case. :D